server/athenian/api/internal/miners/github/deployment.py

```diff
@@ -2025,3 +2028,142 @@ async def load_jira_issues_for_deployments(deployments: pd.DataFrame,
+    )
+
+    if not outlier_deploys:
+        return
+
+    log.info("Clearing %d outlier first deployments", len(outlier_deploys))
+
+    stmts: list[Executable] = []
+
+    # group outlier deploys by name
+    grouped_deploys = [
+        (name, [d.repository_full_name for d in name_groups])
+        for name, name_groups in groupby(
+            sorted(outlier_deploys, key=attrgetter("deployment_name")),
+            key=attrgetter("deployment_name"),
+        )
+    ]
+
+    tables = (GitHubCommitDeployment, GitHubPullRequestDeployment, GitHubReleaseDeployment)
+    for Table in tables:
+        where_clause = sa.and_(
+            Table.acc_id == account,
+            sa.or_(
+                *[
+                    sa.and_(Table.deployment_name == g[0], Table.repository_full_name.in_(g[1]))
+                    for g in grouped_deploys
+                ],
+            ),
+        )
+        stmts.append(sa.delete(Table).where(where_clause))
+
+    # clear GitHubDeploymentFacts by removing prs and commits from facts data
+    FactsT = GitHubDeploymentFacts
+    depl_facts_stmt = sa.select(GitHubDeploymentFacts).where(
+        sa.and_(
+            FactsT.acc_id == account,
+            FactsT.deployment_name.in_(d.deployment_name for d in outlier_deploys),
+        ),
+    )
+    depl_facts_rows = await pdb.fetch_all(depl_facts_stmt)
+
+    for row in depl_facts_rows:
+        facts = DeploymentFacts(row[FactsT.data.name], name=row[FactsT.deployment_name.name])
+        new_facts = facts.with_nothing_deployed()
+        match_cols = (
+            FactsT.acc_id,
+            FactsT.deployment_name,
+            FactsT.release_matches,
+            FactsT.format_version,
+        )
+        update_stmt = sa.update(FactsT).where(
+            sa.and_(*(col == row[col.name] for col in match_cols)),
+        ).values({FactsT.data: new_facts.data})
+        stmts.append(update_stmt)
+
+    await gather(*[pdb.execute(stmt) for stmt in stmts], op="hide_first_deployments")
+
+
+@dataclass(frozen=True, slots=True)
+class _OutlierDeployment:
+    repository_full_name: str
+    deployment_name: str
+
+
+async def _search_outlier_first_deployments(
+    deployment_facts: pd.DataFrame,
+    meta_ids: Sequence[int],
+    mdb: Database,
+    threshold: float = 100.0,
+) -> Sequence[_OutlierDeployment]:
+    """Search the outlier first deployments."""
+    exploded_facts = deployment_facts.explode("repositories").rename(
+        columns={"repositories": "repository"},
+    )
+
+    # retrieve repo => creation time mapping
+    mentioned_repos = exploded_facts.repository.unique()
+    repos_stmt = sa.select(Repository.full_name, Repository.created_at).where(
+        sa.and_(Repository.acc_id.in_(meta_ids), Repository.full_name.in_(mentioned_repos)),
+    )
+    repo_creation_times: Mapping[str, np.datetime64] = {
+        r[0]: np.datetime64(r[1].replace(tzinfo=None)) for r in await mdb.fetch_all(repos_stmt)
```

Comment:
question: is replace(tzinfo=None safe) ? I think asyncpg always return utc, and also in `deployment_facts.started_at` we can only have utc

Reply:
Yes, it should be safe. However, doing this way is very slow. Let's use the built-in numpy conversion for now. I know it is deprecated. I will write a custom Cython function in the future.

Reply:
done